adal/src/Microsoft.IdentityModel.Clients.ActiveDirectory/Exceptions/AdalExceptionFactory.cs

```diff
@@ -40,44 +41,58 @@ namespace Microsoft.IdentityModel.Clients.ActiveDirectory
     internal class AdalExceptionFactory : CoreExceptionFactory
     {
         public override Exception GetClientException(
-            string errorCode, 
-            string errorMessage, 
+            string errorCode,
+            string errorMessage,
             Exception innerException = null)
         {
             ValidateRequiredArgs(errorCode, errorMessage);
             return new AdalException(errorCode, errorMessage);
         }
-      
-        public override Exception GetServiceException(string errorCode, string errorMessage)
+
+        public override Exception GetServiceException(string errorCode, string errorMessage, IHttpWebResponse httpResponse)
```

Comment:
I decided that we are too lax in throwing service exceptions and we throw service exceptions where we should not. So I am making it more complicated, by requiring to have an http response or an exception detail object.

Reply:
This also ensures that we put the headers in the exception.

Reply:
that's a great change. we had a customer issue on this a month or two ago (throwing a service when it should have been client). nice addition.

---
In reply to: [222418975](https://github.com/AzureAD/azure-activedirectory-library-for-dotnet/pull/1294#discussion_r222418975) [](ancestors = 222418975)